### PR TITLE
pass start time to search

### PR DESCRIPTION
**Pull request by borg323**
This implements the idea by @Tilps to pass the start time to the search constructor, so any time lost during setup can be accounted for.

### CODE REVIEW

**Comment by Tilps** [comment]
I was thinking this morning - this might be bad for the nps prediction going in to pruning.  So we might need to keep two start time fields - one for limits and one for nps prediction. (And corresponding two 'GetTimeSinceStart' functions.)

**Comment by borg323** [comment]
When looking at the smart pruning code, [subtracting kSmartPruningToleranceMs from time_since_start](https://github.com/LeelaChessZero/lc0/blob/master/src/mcts/search.cc#L342) seems to be for eliminating such overheads. And this is checked only after 2*kSmartPruningToleranceMs has elapsed, which seems safe.
That being said, having a separate counter for nps cant hurt, so will do it.

File: src/mcts/search.h [hunk 1]
```diff
@@ -167,6 +169,7 @@ class Search {
   Network* const network_;
   const SearchLimits limits_;
   const std::chrono::steady_clock::time_point start_time_;
+  const std::chrono::steady_clock::time_point nps_time_;
```
Comment (Tilps) [inline src/mcts/search.h]: Maybe name this nps_start_time_?

**Comment by Tilps** [comment]
Looks good to me - just the member name could be a bit clearer.

The subtracting and waiting is to offset the 'spin up time' where the first call to the backend can be extraordinarily long compared to all subsequent calls.  I don't think it was intended to offset what we're doing here which is shutting down a previous search.

**Comment by borg323** [comment]
Agreed regarding the purpose of subtracting kSmartPruningToleranceMs. I was wondering whether we could make nps_start_time_ (renamed) also include this 'spin up time'.

File: src/mcts/search.h [hunk 2]
```diff
@@ -57,7 +57,9 @@ class Search {
          BestMoveInfo::Callback best_move_callback,
          ThinkingInfo::Callback info_callback, const SearchLimits& limits,
          const OptionsDict& options, NNCache* cache,
-         SyzygyTablebase* syzygy_tb);
+         SyzygyTablebase* syzygy_tb,
+         std::chrono::steady_clock::time_point start_time =
```
Comment (mooskagh) [inline src/mcts/search.h]: There is a zoo of various start times now...
What do you think about doing this instead:
1. Replace `SearchLimits::time_ms` with `SearchLimits::search_deadline`, make it `std::optional<chrono::steady_clock::time_point>`, and populate in `EngineController::PopulateSearchLimits` (and pass start_time to that function)
2. Make `nps_start_time_` also std::optional and set it after the first iteration which had real NN eval.

Comment (borg323) [reply src/mcts/search.h]: `SearchLimits::time_ms` is also used for selfplay with `--movetime=x` so it isn't easy to remove completely.

Comment (mooskagh) [reply src/mcts/search.h]: There we can also compute end time in game.cc and pass it.

**Comment by mooskagh** [comment]
Which of #404, #427, #450 are intended for merge and in which order?

**Comment by borg323** [comment]
superseded by #450

**Pull request closed without merging**